### PR TITLE
ARROW-13089: [Python] Allow creating RecordBatch from Python dict

### PR DESCRIPTION
**Pull request by kharoc**
Create a from_pydict function in RecordBatch class.
Create unit test for from_pydict

### CODE REVIEW

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -616,6 +616,53 @@ cdef class RecordBatch(_PandasConvertible):
...
+                names.append(k)
+                arrays.append(asarray(v))
+            return RecordBatch.from_arrays(arrays, names, metadata=metadata)
+        elif isinstance(schema, Schema):
+            for field in schema:
+                try:
+                    v = mapping[field.name]
+                except KeyError:
+                    try:
+                        v = mapping[tobytes(field.name)]
```
Comment (pitrou) [inline python/pyarrow/table.pxi:650]: I'm not sure allowing bytes keys is useful.

File: python/pyarrow/tests/test_table.py [hunk 2]
```diff
@@ -1685,3 +1685,60 @@ def test_table_select():
     result = table.select(['f2'])
     expected = pa.table([a2], ['f2'])
     assert result.equals(expected)
+
+
+def test_recordbatch_from_pydict():
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_table.py:1690]: Can you move this next to the existing test for table (to keep the similar tests close to each other). With some parametrization, we might also be able to reduce the duplication (the actual test doesn't seem to rely on anything RecordBatch/Table specific)

**Review by jorisvandenbossche — COMMENTED** [review]
@kharoc thanks for the PR!

Since the implementation is almost identical between Recordbatch and Table, I think we should try to share the code. Both classes don't share a base class at the moment, so maybe a helper function is the easiest (the helper function can return the arguments to be passed to RecordBatch/Table.from_arrays)

File: python/pyarrow/tests/test_table.py (hunk 2 above)
Comment (kharoc) [reply python/pyarrow/tests/test_table.py]: It's solved. Parametrization was added.

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (kharoc) [reply python/pyarrow/table.pxi]: to bytes function was removed.

**Comment by jorisvandenbossche** [comment]
@kharoc can you also take a look at my non-inline comment (https://github.com/apache/arrow/pull/10854#pullrequestreview-721106382) about avoiding duplication between the Table and RecordBatch implementation?

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the update, looks good!

**Pull request closed without merging**